### DISCUSSION TITLE
Security implications of remembering the browser extension credential access for some URL

### DISCUSSION THREAD

**Discussion by jasperweiss** [keepassxreboot/keepassxc#9818] [Q&A] [answered]
When the browser extensions requests access to credentials for a specific URL, KeePassXC prompts you to allow or deny access for specific items. You can remember this decision so it doesn't ask you in the future but I'm wondering if this is safe. The extension is supposedly "untrusted" as it runs in a hostile browser environment. The prompt is from the trusted desktop application itself.
But if you remember this decision, the extension could silently grab the credentials for this URL at any time. Is the URL validated in some way? Could a malicious site like badsite.com manipulate the browser extension to request credentials for mybank.com if access to that item is remembered for that URL? Does the browser itself provide the URL from the page the extension made the request for or does the extension do this? Is this vulnerable to forgery by the site?

It'd be convenient to skip the prompt for access to something like 'MyBank' if I'm on 'https://mybank.com' but I'm just wondering what the security implications would be.

**Comment by droidmonkey** [comment, marked as answer]
> I'm wondering if this is safe

Generally speaking this is up to you and your security requirements.  We think it is safe because we offer the ability to do it. We ask because it is ultimately up to you to decide.

Generally speaking, it isn't possible for a malicious site to spoof the address it is using to the browser. Thus, you are offered some level of assurance against phishing, typo squatting, and malicious embeds (iframes) that use another url than the site you authorized.

**Reply by jasperweiss** [reply]
I've been looking through the extension code and found `sendMessage('retrieve_credentials', [ 'https://somesite.com', kpxc.submitUrl, true ]);` within the context of the content script can fetch arbitrary credentials from the background service part of the extension. My understanding is that content scripts are isolated from scripts from the page itself. So there's no way a script on a site could use this messaging mechanism to fetch credentials from the background service?

Tavis Ormandy [talks a bit about ](https://lock.cmpxchg8b.com/passmgrs.html) content scripts in his blog post on password managers but this seems to be about websites being able to affect the changes the extension makes to the DOM itself such as the icons placed in credential fields in the case of keepassxc. But the content script itself is isolated as far as I understand.
What about the messaging? `chrome.runtime.sendmessage` doesn't work from the webpage context but it does work from say ublock origin's context. Although it doesn't provide a response so presumably it's endpoint is different but I'm wondering if a malicious extension could reach the keepassxc background service somehow.

**Reply by droidmonkey** [reply]
> So there's no way a script on a site could use this messaging mechanism to fetch credentials from the background service?

It's a signal and slot or event driven type setup. When you click the green key icon or request credentials be filled, a request is sent to the background script to do so. All relevant data is pulled from the background script to craft the message back to keepassxc. 

> What about the messaging?

This is more on the browser side of things, but all extensions background portion should be isolated from each other.

**Reply by varjolintu** [reply]
> My understanding is that content scripts are isolated from scripts from the page itself.

This is not entirely true. If a browser extension adds some own content to the page DOM, the page scripts can access that content. Because of this, our extension wraps every content added to DOM in a closed Shadow DOM the page cannot access. It's only accessable using the internal variables in the extension itself, which is isolated from the page and other extensions.

Also, the web page URL is always used for retrieving the credentials. It's not allowed to have a fraudulent iframe and try to retrieve credentials from that. There are some restrictions that disallow running content scripts from other domains.

> Although it doesn't provide a response so presumably it's endpoint is different but I'm wondering if a malicious extension could reach the keepassxc background service somehow.

In theory, if there's a fraudulent application in your computer, it could try to access your database using the Brower Extension API and ask for a URL that's automatically accepted. This would also require spoofing the Connection ID and its generated key, so it's not as simple as "just asking" for it. Those who are worried about 3rd party applications accessing the API locally, this change will reduce the possible attack surface a bit: https://github.com/keepassxreboot/keepassxc/pull/9779

Basically, the feature is added for convenience, although it's preferred to have a confirmation that requires user interaction.

**Reply by jasperweiss** [reply]
> In theory, if there's a fraudulent application in your computer, it could try to access your database using the Brower Extension API and ask for a URL that's automatically accepted.

I understand the extensions does some form of a key exchange during setup with presumably a form of TOFU key verification (which is always confirmed explicitly in the application prompt). I wasn't referring to the communication of the extension itself with the application through the socket connection but rather the communication of the conent script with the background service script through the messaging system. 

See [this](https://github.com/keepassxreboot/keepassxc-browser/blob/f3379e6677f9ab65e68387fe5bb883059d30b6e9/keepassxc-browser/content/keepassxc-browser.js#L629C90-L629C90) line. It uses a messaging system to retrieve credentials for some arbitrary url from the background service, which then requests them from the application over the secured socket connection. I can run that same function within the browser terminal from the KeePassXC extension's context and request credentials for arbitrary url's (obviously resulting in a confirmation prompt from the application unless access is remembered). So I was wondering whether this messaging system between the content script and the background service script is sufficiently shielded? Could other extension run `sendMessage('retrieve_credentials', [ kpxc.url, kpxc.submitUrl, true ]` somehow? I don't understand this messaging system fully so that's why I'm asking.

**Reply by varjolintu** [reply]
> So I was wondering whether this messaging system between the content script and the background service script is sufficiently shielded? Could other extension run sendMessage('retrieve_credentials', [ kpxc.url, kpxc.submitUrl, true ] somehow? I don't understand this messaging system fully so that's why I'm asking.

It could not. Another extension does not have access to other extension's contents. Another thing that extension content scripts rely on is `isTrusted()` that is `true` only if the action was actually manually done by the user. A script, even if it somehow could have access to the extension's scripts, could not run an operation automatically.

You could test this by yourself downloading some example extension and running that command inside it when you load it as a temporary extension.

**Comment by michaelk83** [comment]
Regardless of the technical considerations that are explained by droidmonkey, if you want to be a bit more careful, and have more peace of mind, you can skip the access prompts for low-risk stuff like blogs and forums, and leave the prompts on for high-risk stuff like your bank account.
